### DISCUSSION TITLE
Does disposing a pipeline registry dispose chained limiters?

### DISCUSSION THREAD

**Discussion by lonix1** [App-vNext/Polly#2425] [Q&A]
One benefit of the "resilience registry" is that it is easy to [dispose](https://www.pollydocs.org/pipelines/resilience-pipeline-registry.html#resource-disposal):
> The registry caches and manages all pipelines and resources linked to them. When you dispose of the registry, all pipelines created by it are also disposed of and can't be used anymore.

Does that also include chained rate limiters?

Suppose I have a pipeline with multiple limiters (e.g. concurrency limiter and rate limiter) and also a chained rate limiter which itself contains rate limiters - are all of those disposed when the registry is disposed?

**Comment by migig** [comment]
@lonix1 Hi, do you already have this knowledge? I need this solution also.

**Comment by peter-csala** [comment]
I think there is some misunderstanding so, let me try to clarify.

## Registry and Pipeline
So, whenever you register a pipeline inside a registry then you can't dispose the pipeline directly. (The `DisposeHelper` is `internal`.)

https://github.com/App-vNext/Polly/blob/4f0d44a1fb7b984afe27a6b51fb63940876b1e3f/test/Polly.Core.Tests/Registry/ResiliencePipelineRegistryTests.cs#L439-L446

You can only dispose the pipeline through the registry:

https://github.com/App-vNext/Polly/blob/4f0d44a1fb7b984afe27a6b51fb63940876b1e3f/test/Polly.Core.Tests/Registry/ResiliencePipelineRegistryTests.cs#L410-L436

## Disposed Pipeline

After the registry has been disposed you obviously can't retrieve a pipeline because you will get an `ObjectDisposedException`. 

Let's suppose you have retrieved a pipeline just before the registry disposal. If you want to invoke a method on the pipeline like the `Execute{Async}` then you will receive an `ObjectDisposedException`.

Retrieving a strategy from a pipeline can be achieved via `Polly.Testing`. But the strategies are defined as `internal`. Also the `ResilienceStrategy`'s `ExecuteCore` is `internal`. But with a bit of reflection magic it can be called

```cs
var registry = new ResiliencePipelineRegistry<string>();
registry.TryAddBuilder("A", (builder, _) =>
{
	builder.AddRateLimiter(new SlidingWindowRateLimiter(
		new SlidingWindowRateLimiterOptions
		{
			PermitLimit = 10,
			Window = TimeSpan.FromMinutes(1),
			SegmentsPerWindow = 1
		}));
});

var pipeline = registry.GetPipeline("A");
registry.Dispose();

try
{
	pipeline.Execute(() => {});
}
catch(ObjectDisposedException)
{
	Console.WriteLine("Pipeline has been disposed");
}

var strategy = pipeline.GetPipelineDescriptor().FirstStrategy.StrategyInstance;
var executeCore = typeof(ResilienceStrategy).GetMethod("ExecuteCore", BindingFlags.Instance | BindingFlags.NonPublic);
executeCore = executeCore.MakeGenericMethod(new[] {typeof(object), typeof(object)});

var context = ResilienceContextPool.Shared.Get();
Func<ResilienceContext, object, ValueTask<Outcome<object>>> callback = (_, __) => Outcome.FromResultAsValueTask<object>("Test");
var executeTask = (ValueTask<Outcome<object>>)executeCore.Invoke(strategy, parameters: [callback, context, null]);
var executeOutcome = await executeTask;
ResilienceContextPool.Shared.Return(context);

Console.WriteLine(executeOutcome.Result);
```
It will print 
```
Pipeline has been disposed
Test
```
So, as you can see the pipeline is disposed but the strategy isn't.

[Dotnet Fiddle](https://dotnetfiddle.net/eNoLF7)

**Comment by lonix1** [comment]
Happy New Year @peter-csala. Thanks for that detail, it's helpful.

To be sure I understand you, here is a simplified example:
```cs
_resiliencePipelineRegistry.TryAddBuilder("A", (builder, context) => {

  builder.AddRateLimiter(new RateLimiterStrategyOptions {
    RateLimiter = args => CreateRateLimiter(...).AcquireAsync(...)                         // (1)
  });

  builder.AddRateLimiter(new RateLimiterStrategyOptions {
    RateLimiter = args => PartitionedRateLimiter.CreateChained(...).AcquireAsync(...)      // (2)
  });

});
```

I'm using [IDisposableAnalyzers](https://github.com/DotNetAnalyzers/IDisposableAnalyzers) to detect undisposed resources. It warns me that both `(1)` and `(2)` should be disposed.

Will those rate limiters (even a "chained" one which contains other limiters) be disposed by the registry itself?

**Reply by migig** [reply]
@lonix1 Thank you for code which I will study.

This message actually regarding https://github.com/DotNetAnalyzers/IDisposableAnalyzers. I try on large project and find many bugs for me - thank you!

**Comment by peter-csala** [comment]
Happy New Year @lonix1  you too!

## Polly.Core
Let's see what happens when you call the `Dispose{Async}` on the registry.

1. The `ResiliencePipelineRegistry`'s` Dispose` calls the `ResiliencePipeline`'s `DisposeHelper.ForceDisposeAsync()`
2. The `ResiliencePipeline`'s `DisposeHelper` is a `ComponentDisposeHelper`
3. The `ComponentDisposeHelper` wraps the component and exposes several `Dispose` related helper methods
4. The `ForceDisposeAsync` calls the `DisposeAsync` on component
5. The component is a `PipelineComponent` which defines the `DisposeAsync` as abstract
6. For example the `CompositeComponent`'s `DisposeAsync` calls all its components' `DisposeAsync`

As you can see on the following image the `PipelineComponent` is a strategy (in this particular case a Timeout)
![polly core](https://github.com/user-attachments/assets/34337acf-597c-4a53-8da5-144a70b303bc)

So, if the given strategy is disposable then it will be disposed. BUT not all strategies are disposable. For example the `CircuitBreakerResilienceStrategy` is disposable, but the `TimeoutResilienceStrategy` isn't.

## Polly.RateLimiting

The things are bit tricky here.

1. The `RateLimiterResilienceStrategy` has a property called `Wrapper`
2. This `Wrapper` is a `DisposeWrapper`
3. The strategy's `DisposeAsync` calls the wrapper's `DisposeAsync`
4. The `DisposeWrapper`'s has a reference to the `RateLimiter` and its `DisposeAsync` calls the limiter's `DisposeAsync`

BUT inside the `RateLimiterResiliencePipelineBuilderExtensions` the most generic `AddRateLimiter` has a conditional logic.

https://github.com/App-vNext/Polly/blob/4f0d44a1fb7b984afe27a6b51fb63940876b1e3f/src/Polly.RateLimiting/RateLimiterResiliencePipelineBuilderExtensions.cs#L101-L129

If the `RateLimiter` property of the `options` is defined then it will create a `RateLimiterResilienceStrategy` with a `null` as a `DisposeWrapper`.

So, your limiter won't be disposed because the underlying strategies `DisposeWrapper` are `null`. 

![polly ratelimiter](https://github.com/user-attachments/assets/01161a50-cc61-4d7a-9e99-387928858094)

**Reply by peter-csala** [reply]
@martintmk It seems to me that we have found a bug. If the `options.RateLimiter` is defined then we should say `wrapper = new DisposeWrapper(options.RateLimiter);` Or did I miss something?

**Reply by martintmk** [reply]
> It seems to me that we have found a bug. If the options.RateLimiter is defined then we should say wrapper = new DisposeWrapper(options.RateLimiter); Or did I miss something?

The problem is that `options.RateLimiter` is just a callback and it cannot be disposed. The `DisposeWrapper` is only ever used if no custom rate limiter is defined, then we know that we are creating the default, explicit instance of `ConcurrencyLimiter` that is disposable. 


If one is using custom limiters and want to dispose then on pipeline reload or when registry is disposed, they can use `OnPipelineDisposed` [callback](https://www.pollydocs.org/api/Polly.Registry.ConfigureBuilderContext-1.html#Polly_Registry_ConfigureBuilderContext_1_OnPipelineDisposed_System_Action_).

This is explained in our docs:
https://www.pollydocs.org/strategies/rate-limiter.html#disposal-of-rate-limiters

**Reply by peter-csala** [reply]
Ohh I've missed that part. Thanks @martintmk

**Reply by lonix1** [reply]
Thanks @peter-csala and @martintmk! It was a lot to digest, but I think I understand it now.

I created a [runnable minimal example](https://dotnetfiddle.net/HOKUvR) showing multiple limiters and their disposal. It's generic enough as an example for other users in the future.

Here is the code for reference:
```cs
using System;
using System.Threading.RateLimiting;
using Polly;
using Polly.RateLimiting;
using Polly.Registry;


public static class Program
{	
  public static void Main()
  {
    using var pra = new PipelineRegistryAdapter();
    pra.GetOrCreateResiliencePipeline("Pipeline foo", 1, 10, 100, 1000);
    pra.GetOrCreateResiliencePipeline("Pipeline bar", 2, 20, 200, 2000);
  }
}


public sealed class PipelineRegistryAdapter : IDisposable
{
  private bool _disposed;
  private readonly ResiliencePipelineRegistry<string> _resiliencePipelineRegistry = new();


  public void Dispose()
  {
    if (_disposed) return;
    _resiliencePipelineRegistry.Dispose();
    _disposed = true;
  }


  private PartitionedRateLimiter<ResilienceContext> CreateConcurrencyLimiter(string partitionKey, int permitLimit) =>
    PartitionedRateLimiter.Create<ResilienceContext, string>(context =>
      RateLimitPartition.GetConcurrencyLimiter(
        partitionKey: partitionKey,
        factory:      partitionKey => new ConcurrencyLimiterOptions { PermitLimit = permitLimit, QueueLimit = 0 }));


  private PartitionedRateLimiter<ResilienceContext> CreateFixedWindowLimiter(string partitionKey, int permitLimit, TimeSpan window) =>
    PartitionedRateLimiter.Create<ResilienceContext, string>(context =>
      RateLimitPartition.GetFixedWindowLimiter(
        partitionKey: partitionKey,
        factory:      partitionKey => new FixedWindowRateLimiterOptions { PermitLimit = permitLimit, QueueLimit = 0, Window = window }));


  public ResiliencePipeline GetOrCreateResiliencePipeline(string partitionKey, int maximumConcurrentThreads, int sendLimitPerSecond, int sendLimitPerHour, int sendLimitPerDay)
  {
    // return pipeline if exists
    if (_resiliencePipelineRegistry.TryGetPipeline(partitionKey, out var pipeline))
      return pipeline;

    // else create pipeline with multiple strategies
    var wasCreated = _resiliencePipelineRegistry.TryAddBuilder(partitionKey, (builder, context) =>
    {
      PartitionedRateLimiter<ResilienceContext>? threadLimiter = null;
      PartitionedRateLimiter<ResilienceContext>? requestLimiter = null;

      // outer strategy: limit threads
      builder.AddRateLimiter(new RateLimiterStrategyOptions {
        RateLimiter = args => {
          threadLimiter = CreateConcurrencyLimiter(partitionKey, maximumConcurrentThreads);
          return threadLimiter.AcquireAsync(args.Context, permitCount:1, args.Context.CancellationToken);
        }
      });

      // inner strategy: limit requests (by second, hour, day)
      builder.AddRateLimiter(new RateLimiterStrategyOptions {
        RateLimiter = args => {
          PartitionedRateLimiter<ResilienceContext>[] limiters = [
            CreateFixedWindowLimiter(partitionKey, sendLimitPerSecond, TimeSpan.FromSeconds(1)),
            CreateFixedWindowLimiter(partitionKey, sendLimitPerHour,   TimeSpan.FromHours(1)),
            CreateFixedWindowLimiter(partitionKey, sendLimitPerDay,    TimeSpan.FromDays(1)),
          ];
          requestLimiter = PartitionedRateLimiter.CreateChained(limiters);
          return requestLimiter.AcquireAsync(args.Context, permitCount:1, args.Context.CancellationToken);
        }
      });

      // unlike other strategies, rate limiters disposed manually         // <---------- DISPOSAL HERE
      context.OnPipelineDisposed(() => {
        threadLimiter?.Dispose();
        requestLimiter?.Dispose();
        Console.WriteLine($"Disposed pipeline '{partitionKey}'");
      });

    });

    if (wasCreated)
      Console.WriteLine($"Created pipeline '{partitionKey}'");
    else
      throw new InvalidOperationException($"Failed to create pipeline '{partitionKey}'");

    return _resiliencePipelineRegistry.GetPipeline(partitionKey);
  }


}
```

It uses the `OnPipelineDisposed` callback to dispose the limiters.

It stores limiters in closures which can be accessed by the callback; I dislike that coding style in C#, but it's fine here.

For someone who is not a "Polly insider", this code is complicated... but seems to work. I was wondering if the disposal part "looks OK"? I dispose the "chained" limiter only (`requestLimiter`), instead of its child limiters, is that OK?

**Reply by peter-csala** [reply]
Hey @lonix1 

Whenever you call `CreateChained` then a `ChainedPartionedRateLimiter` will be created. 

Based on its `Dispose` method it disposes the limiters in a reverse order:

https://github.com/dotnet/runtime/blob/3bbd4037e17e2fc4701a286f3eee18bbbadc8c67/src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ChainedPartitionedRateLimiter.cs#L227-L257

> // Dispose in reverse order
> // Avoids issues where dispose might unblock a queued acquire and then the acquire fails when acquiring the next limiter.
> // When disposing in reverse order there wont be any issues of unblocking an acquire that affects acquires on limiters in the chain after it

**Reply by lonix1** [reply]
Thanks! I forgot the rate limiting is in the framework, rather than Polly.

I'm glad it's finally working. Thanks guys for your advice. :smile: 

*BTW: that example is as simple as I could make it. One could use it when other people ask questions.*

**Reply by peter-csala** [reply]
If you think that it's worth sharing with others then please consider to contribute into the Polly docs. 

If you need some help with that please let us know.

**Reply by lonix1** [reply]
If you think it'll add value, I could do that. It could be a new subsection under [registry - resource disposal](https://www.pollydocs.org/pipelines/resilience-pipeline-registry.html#resource-disposal).

I'm unsure though. Like I mentioned, I made that example as "simple as possible", but in a previous comment also mentioned that it was very complicated for me (as an outsider). If we put that sample app into the docs it may confuse many people unnecessarily?

**Reply by lonix1** [reply]
Actually there are two ways:
- Put that code directly into the docs; maybe confusing as it's a complicated topic
- Or, explain in a small paragraph that rate limiters are the one kind of strategy that should be disposed manually. And link to a fiddle with the full code, or maybe add link to this discussion.

I could do either, what do you think? (But you'd need to review that code in case I made a mistake... It works for me, but I'm not a Polly expert.)

**Reply by peter-csala** [reply]
If you want I can try to simplify your sample code next week.

**Reply by lonix1** [reply]
I submitted a [PR](https://github.com/App-vNext/Polly/pull/2438) with the full example. Let me know how to tweak it. No rush!
